### DISCUSSION TITLE
Integrating Twenty's Core Business Logic into a Monolithic Architecture with Custom Components

### DISCUSSION THREAD

**Discussion by thefrana** [twentyhq/twenty#8605] [Ideas]
Dear Twenty Team,

First and foremost, thank you for developing such a robust and flexible open-source CRM solution. The features and modularity of Twenty are impressive, and we are keen to integrate it into our existing systems.

**Our Context:**

We are a software company operating a monolithic architecture and are interested in leveraging Twenty's core business logic while customizing other components to fit our specific requirements. Our objectives include:

- Utilizing Twenty's Core Business Logic: We aim to integrate Twenty's CRM functionalities directly into our monolithic application, rather than deploying it as a standalone service.
- Implementing a Custom SQL Database: We plan to connect Twenty's core logic to our existing custom SQL database schema, differing from the default implementation.
- Developing Custom Admin Components: We intend to create a tailored admin interface that aligns with our current system, potentially reusing some of Twenty's components where feasible.
- Establishing an Adapter Layer: To facilitate this integration, we anticipate the need for an adapter layer between Twenty and our monolithic architecture.

**Our Questions:**

1. Is it feasible to decouple Twenty's business logic and integrate it into a monolithic application?
1. Can we override or extend Twenty's modules to utilize a custom SQL database while maintaining the integrity of the core logic?
1. Are there best practices or existing tools within the Twenty ecosystem that would assist in adapting it to a custom admin interface?
1. What potential challenges or limitations should we be aware of when pursuing this integration approach?

We are eager to explore how we can achieve this integration while adhering to Twenty's design principles. Any guidance, documentation, or examples you could provide would be greatly appreciated.

Thank you for your time and assistance.

Lukas

**Comment by tony9823** [comment]
Could you provide detailed documentation or guidance on customizing the integration of Twenty with our existing systems, particularly in terms of APIs, data migration, and ensuring compatibility with our current tech stack?

**Reply by thefrana** [reply]
I am not sure this is a question for me. Or is it?

**Comment by thatmike1** [comment]
Hey @thefrana , I stumbled across this thread while researching Twenty for a similar (though not identical) use case.

We're evaluating Twenty as a foundation for a construction industry CRM/ERP - basically extending it with 8+ custom modules (HR, invoicing, inventory, site diaries, etc.) rather than embedding it into an existing app like you described.

Curious if you ever made progress on your approach? Specifically:

- Did you find a clean way to leverage Twenty's business logic/custom objects system?
- Any gotchas or limitations you hit that aren't obvious from the docs?
- Did you end up going a different direction?

Any lessons learned would be super helpful - there's not much documented about using Twenty for larger-scale custom work. Thanks!
PS: Happy to share what we learn from our PoC if it helps others in the same boat :)

**Reply by thefrana** [reply]
Hi, as the agentic universe is expanding, we eventually found that a better solution is to let agents rewrite the Twenty codebase into our custom codebase instead of making an adapter. It can derive the database structure (we use a different adapter, but rewriting the schema is quite easy) and components (we have a custom UI library). The only challenging thing is the business logic, but even that is becoming more transferable. It went really well.
